src/main/java/Duke.java

```diff
@@ -1,10 +1,186 @@
+    }
+
+
+    //add function for level-2
+    public static void add(String task) {
+        Task newTask = new Task(task);
+        taskList[taskCount] = newTask;
+        taskCount++;
+        System.out.println("added: " + task);
+    }
+
+
+    //list function for level-2
+    public static void list() {
+        for (int i = 0; i < taskCount; i++) {
+            System.out.print((i + 1) + ". ");
+            System.out.println(taskList[i].toString());
+        }
+    }
+
+
+    //mark function for level-3
+    public void mark(int index) {
+        if (index > taskCount || index < 0) {
+            System.out.println("Index out of bounds");
+            return;
+        }
+        System.out.println("Nice! I've marked this task as done:");
+        taskList[index - 1].markAsDone();
+    }
+
+
+    //unmark function for level-3
+    public void unmark(int index) {
+        if (index > taskCount || index < 0) {
+            System.out.println("Index out of bounds");
+            return;
+        }
+        System.out.println("Ok, I've marked this task as not done yet:");
+        taskList[index - 1].markAsNotDone();
+
+    }
+
+
+    //add toDo task for level-4
+    public static void addToDo(String[] inputComponent) {
+        String description = "";
+
+        for (int i = 1; i < inputComponent.length; i++) {
+            description += inputComponent[i];
+            description += " ";
+        }
+
+        taskList[taskCount] = new ToDo(description.trim());
+        System.out.println("Got it. I've added this task:");
+        System.out.println(taskList[taskCount].toString());
+        taskCount++;
+        System.out.println("Now you have " + taskCount + " tasks in the list");
+    }
+
+
+    //add deadline task for level-4
+    public static void addDeadline(String[] inputComponent) {
+        String description = "";
+        String by = "";
+        int state = 0;//transition from "description" to "by" string
+
+
+        for (int i = 1; i < inputComponent.length; i++) {
+            if (inputComponent[i].equals("/by")) {
+                state = 1;
+            } else {
+                if (state == 1) {
+                    by += inputComponent[i];
+                    by += " ";
+                } else {
+                    description += inputComponent[i];
+                    description += " ";
+                }
+            }
+        }
+
+
+        taskList[taskCount] = new Deadline(description.trim(), by.trim());
+        System.out.println("Got it. I've added this task:");
+        System.out.println(taskList[taskCount].toString());
+        taskCount++;
+        System.out.println("Now you have " + taskCount + " tasks in the list");
+    }
+
+
+    //add event task for level-4
+    public static void addEvent(String[] inputComponent) {
+        String description = "";
+        String from = "";
+        String to = "";
+        int state = 0;//transition from "description" to "from" to "to" string
```

Comment:
Should `state` here be named a bit better to give readers and future developers an indication of what it does instead? I don't like how it is vague as of now. Since `state` can only take values from `[0, 1, 2]`, it might also be worth it to create an `enum` and some constants to indicate what each state means, to get rid of "magic numbers".